src/ILCompiler.Compiler/src/Compiler/ExportedMethodsRootProvider.cs

```diff
@@ -36,11 +45,45 @@ public void AddCompilationRoots(IRootingServiceProvider rootProvider)
                     if (ecmaMethod.IsNativeCallable)
                     {
                         string nativeCallableExportName = ecmaMethod.GetNativeCallableExportName();
+
                         if (nativeCallableExportName != null)
+                        {
+                            _methods.Add(ecmaMethod);
                             rootProvider.AddCompilationRoot(method, "Native callable", nativeCallableExportName);
+                        }
                     }
                 }
             }
+
+            // Only emit exported methods for user supplied assemblies
+            // _exportsFile is null when we're not doing a shared library build
+            if (_module != _module.Context.SystemModule && _exportsFile != null)
+                EmitExportedMethods();
+        }
+
+        private void EmitExportedMethods()
+        {
+            string moduleName = Path.GetFileNameWithoutExtension(_exportsFile);
+            StringBuilder stringBuilder = new StringBuilder();
+
+            if (_module.Context.Target.IsWindows)
+            {
+                stringBuilder.Append("LIBRARY   ");
+                stringBuilder.AppendLine(moduleName.ToUpper());
+
+                stringBuilder.AppendLine("EXPORTS");
+                foreach (var method in _methods)
+                    stringBuilder.AppendLine("   " + method.GetNativeCallableExportName());
```

Comment:
Nit: You can consider using interpolated strings for all these - it looks better:

`AppendLine($"   {method.GetNativeCallableExportName()}");`